File: core/src/main/java/vn/loitp/views/shadowviewhelper/ShadowViewDrawable.java
package vn.loitp.views.shadowviewhelper;

import android.graphics.Canvas;
import android.graphics.ColorFilter;
import android.graphics.Paint;
import android.graphics.PixelFormat;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.Rect;
import android.graphics.RectF;
import android.graphics.drawable.Drawable;

import androidx.annotation.NonNull;

/**
 * <code>
 * if (Build.VERSION.SDK_INT > Build.VERSION_CODES.HONEYCOMB) {
 * view.setLayerType(View.LAYER_TYPE_SOFTWARE, drawable.getPaint());
 * }
 * </code>
 * <p/>
 * Author: wangjie
 * Email: tiantian.china.2@gmail.com
 * Date: 5/2/15.
 */
public class ShadowViewDrawable extends Drawable {
    private Paint paint;
    private RectF bounds = new RectF();
    private int width;
    private int height;
    private ShadowProperty shadowProperty;
    private int shadowOffset;
    private RectF drawRect;
    private float rx;
    private float ry;

    public ShadowViewDrawable(ShadowProperty shadowProperty, int color, float rx, float ry) {
        this.shadowProperty = shadowProperty;
        shadowOffset = this.shadowProperty.getShadowOffset();

        this.rx = rx;
        this.ry = ry;

        paint = new Paint();
        paint.setAntiAlias(true);
        paint.setFilterBitmap(true);
        paint.setDither(true);
        paint.setStyle(Paint.Style.FILL);
        paint.setColor(color);
        paint.setShadowLayer(shadowProperty.getShadowRadius(), shadowProperty.getShadowDx(), shadowProperty.getShadowDy(), shadowProperty.getShadowColor());

        drawRect = new RectF();
    }

    @Override
    protected void onBoundsChange(Rect bounds) {
        super.onBoundsChange(bounds);
        if (bounds.right - bounds.left > 0 && bounds.bottom - bounds.top > 0) {
            this.bounds.left = bounds.left;
            this.bounds.right = bounds.right;
            this.bounds.top = bounds.top;
            this.bounds.bottom = bounds.bottom;
            width = (int) (this.bounds.right - this.bounds.left);
            height = (int) (this.bounds.bottom - this.bounds.top);

//            drawRect = new RectF(shadowOffset, shadowOffset, width - shadowOffset, height - shadowOffset);
//            drawRect = new RectF(0, 0, width, height - shadowOffset);

            final int shadowSide = shadowProperty.getShadowSide();
            final int left = (shadowSide & ShadowProperty.Companion.getLEFT()) == ShadowProperty.Companion.getLEFT() ? shadowOffset : 0;
            final int top = (shadowSide & ShadowProperty.Companion.getTOP()) == ShadowProperty.Companion.getTOP() ? shadowOffset : 0;
            final int right = width - ((shadowSide & ShadowProperty.Companion.getRIGHT()) == ShadowProperty.Companion.getRIGHT() ? shadowOffset : 0);
            final int bottom = height - ((shadowSide & ShadowProperty.Companion.getBOTTOM()) == ShadowProperty.Companion.getBOTTOM() ? shadowOffset : 0);

            drawRect = new RectF(left, top, right, bottom);

            invalidateSelf();
        }
    }

    private PorterDuffXfermode srcOut = new PorterDuffXfermode(PorterDuff.Mode.SRC_OUT);

    @Override
    public void draw(@NonNull Canvas canvas) {
        paint.setXfermode(null);

        canvas.drawRoundRect(
                drawRect,
                rx, ry,
                paint
        );

        paint.setXfermode(srcOut);
//        paint.setColor(Color.TRANSPARENT);
        canvas.drawRoundRect(drawRect, rx, ry, paint);
    }

    public ShadowViewDrawable setColor(int color) {
        paint.setColor(color);
        return this;
    }

    @Override
    public void setAlpha(int alpha) {

    }

    @Override
    public void setColorFilter(ColorFilter cf) {

    }

    @Override
    public int getOpacity() {
        return PixelFormat.UNKNOWN;
    }
}

File: core/src/main/java/com/loitp/views/shadowViewHelper/ShadowViewDrawable.kt
package com.loitp.views.shadowViewHelper

import android.graphics.* // ktlint-disable no-wildcard-imports
import android.graphics.drawable.Drawable

/**
 * Created by Loitp on 04,August,2022
 * Galaxy One company,
 * Vietnam
 * +840766040293
 * freuss47@gmail.com
 */
class ShadowViewDrawable(
    private val shadowProperty: ShadowProperty,
    color: Int,
    private val rx: Float,
    private val ry: Float
) : Drawable() {
    private val paint: Paint = Paint()
    private val bounds = RectF()
    private var width: Int = 0
    private var height: Int = 0
    private val shadowOffset: Int = this.shadowProperty.shadowOffset
    private var drawRect: RectF = RectF()

    private val srcOut = PorterDuffXfermode(PorterDuff.Mode.SRC_OUT)

    init {
        paint.isAntiAlias = true
        paint.isFilterBitmap = true
        paint.isDither = true
        paint.style = Paint.Style.FILL
        paint.color = color
        paint.setShadowLayer(
            shadowProperty.getShadowRadius().toFloat(),
            shadowProperty.getShadowDx().toFloat(), shadowProperty.getShadowDy().toFloat(),
            shadowProperty.getShadowColor()
        )
    }

    override fun onBoundsChange(bounds: Rect) {
        super.onBoundsChange(bounds)
        if (bounds.right - bounds.left > 0 && bounds.bottom - bounds.top > 0) {
            this.bounds.left = bounds.left.toFloat()
            this.bounds.right = bounds.right.toFloat()
            this.bounds.top = bounds.top.toFloat()
            this.bounds.bottom = bounds.bottom.toFloat()
            width = (this.bounds.right - this.bounds.left).toInt()
            height = (this.bounds.bottom - this.bounds.top).toInt()

            //            drawRect = new RectF(shadowOffset, shadowOffset, width - shadowOffset, height - shadowOffset);
            //            drawRect = new RectF(0, 0, width, height - shadowOffset);

            val shadowSide = shadowProperty.getShadowSide()
            val left =
                if (shadowSide and ShadowProperty.LEFT == ShadowProperty.LEFT) shadowOffset else 0
            val top =
                if (shadowSide and ShadowProperty.TOP == ShadowProperty.TOP) shadowOffset else 0
            val right =
                width - if (shadowSide and ShadowProperty.RIGHT == ShadowProperty.RIGHT) shadowOffset else 0
            val bottom =
                height - if (shadowSide and ShadowProperty.BOTTOM == ShadowProperty.BOTTOM) shadowOffset else 0

            drawRect = RectF(left.toFloat(), top.toFloat(), right.toFloat(), bottom.toFloat())

            invalidateSelf()
        }
    }

    override fun draw(canvas: Canvas) {
        paint.xfermode = null

        canvas.drawRoundRect(
            drawRect,
            rx, ry,
            paint
        )

        paint.xfermode = srcOut
        //        paint.setColor(Color.TRANSPARENT);
        canvas.drawRoundRect(drawRect, rx, ry, paint)
    }

    fun setColor(color: Int): ShadowViewDrawable {
        paint.color = color
        return this
    }

    override fun setAlpha(alpha: Int) {
    }

    override fun setColorFilter(cf: ColorFilter?) {
    }

    override fun getOpacity(): Int {
        return PixelFormat.UNKNOWN
    }
}
